### DISCUSSION TITLE
Python path modification when using pytest directly

### DISCUSSION THREAD

**Discussion by pmeier** [pytest-dev/pytest#10902] [Q&A]
Imagine I have a conda env like this

1. `conda create -n pytest-q python=3.8`
2. `conda activate pytest-q`
3. `pip install pytest`

If I look at the Python path now, it looks like this:

```
$ python -c 'import sys; print("\n".join(sys.path))'
/home/philip/.conda/envs/pytest-q/lib/python38.zip
/home/philip/.conda/envs/pytest-q/lib/python3.8
/home/philip/.conda/envs/pytest-q/lib/python3.8/lib-dynload
/home/philip/.conda/envs/pytest-q/lib/python3.8/site-packages
```

Imagine further, I have a `test.py` module like this:

```py
import sys


def test():
    print("\n".join(sys.path))
```

Running `pytest` through Python gives this result

```
$ echo $PWD
/home/philip/git/nebari-dev/nebari
$ python -m pytest -sq test.py
/home/philip/git/nebari-dev/nebari
/home/philip/.conda/envs/pytest-q/lib/python38.zip
/home/philip/.conda/envs/pytest-q/lib/python3.8
/home/philip/.conda/envs/pytest-q/lib/python3.8/lib-dynload
/home/philip/.conda/envs/pytest-q/lib/python3.8/site-packages
.
1 passed in 0.00s
```

As explained [here](https://docs.pytest.org/en/7.3.x/explanation/pythonpath.html#invoking-pytest-versus-python-m-pytest), it is standard Python behavior to include the current working directory. Meaning everything is good here.

However, running `pytest` directly yields

```
$ pytest -sq test.py
/home/philip/git/nebari-dev/nebari
/home/philip/.conda/envs/pytest-q/bin
/home/philip/.conda/envs/pytest-q/lib/python38.zip
/home/philip/.conda/envs/pytest-q/lib/python3.8
/home/philip/.conda/envs/pytest-q/lib/python3.8/lib-dynload
/home/philip/.conda/envs/pytest-q/lib/python3.8/site-packages
```

So not only was the current working directory added as well, the binary directory of my conda env, i.e. `$CONDA_PREFIX/bin`, was added as well.

Is this expected? I couldn't find any documentation on this.

Edit:

Ok, this has nothing to do with `conda`. Can reproduce the same behavior inside a [`virtualenv`](https://pypi.org/project/virtualenv/).

**Comment by RonnyPfannschmidt** [comment]
that seems like a bug, whats the operating system and where is the python from

**Reply by pmeier** [reply]
```
❯ uname -o
GNU/Linux
❯ python -VV
Python 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:01:55) 
[GCC 11.3.0]
```

**Reply by pmeier** [reply]
Can also reproduce with Python from `conda` defaults channel:

```
❯ python -VV
Python 3.8.16 (default, Mar  2 2023, 03:21:46) 
[GCC 11.2.0]
❯ pytest -V
pytest 7.3.0
```

`pytest` version was the same in all setups.

**Reply by pmeier** [reply]
Can trace it back to at least `pytest==5.4.3`. Meaning, in case this is actually a bug, it is not a recent regression.

**Reply by RonnyPfannschmidt** [reply]
So the key question is how the paths come to be for those setups

So these happen to non conda python interpreters

**Reply by pmeier** [reply]
Yes, this does seem to happen on non-conda Python binaries as well.

```
$ nix-shell -p python38Packages.pytest
$ python -VV
Python 3.8.16 (default, Dec  6 2022, 19:12:06) 
[GCC 11.3.0]
$ pytest -V
pytest 7.1.3
$ python -m pytest -sq test.py | head -n3
/home/philip
/nix/store/25zarrb1q1wpin34vh20am0hlfzkv0rc-python3.8-pytest-7.1.3/lib/python3.8/site-packages
/nix/store/0jcddjpn4wnx6pyr6hlmxw3xzgszqm81-python3.8-attrs-22.1.0/lib/python3.8/site-packages
$ pytest -sq test.py | head -n3
/home/philip
/nix/store/25zarrb1q1wpin34vh20am0hlfzkv0rc-python3.8-pytest-7.1.3/bin
/nix/store/25zarrb1q1wpin34vh20am0hlfzkv0rc-python3.8-pytest-7.1.3/lib/python3.8/site-packages
```

**Reply by pmeier** [reply]
Want me to open a proper issue for this?

**Reply by RonnyPfannschmidt** [reply]
Does it happen on non-nix pythons, nix has numerous hacks in place that may play a role

**Reply by pmeier** [reply]
The last option that I have available is an Arch box. It seems the behavior is the same:

```
❯ python -VV
Python 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201]
❯ pytest -V
pytest 7.3.0
❯ python -m pytest -sq test.py | head -n3
/home/philip
/usr/lib/python310.zip
/usr/lib/python3.10
❯ pytest -sq test.py | head -n3
/home/philip
/usr/bin
/usr/lib/python310.zip
```

**Comment by nicoddemus** [comment]
I get similar results on Windows:

(cwd is `c:\Users\bruno\projects\pytest`)

```
λ python -m pytest -sq .tmp\test_p.py
c:\Users\bruno\projects\pytest\.tmp
c:\Users\bruno\projects\pytest
C:\Users\bruno\AppData\Local\Programs\Python\Python310\python310.zip
...
```

```
λ pytest -sq .tmp\test_p.py
c:\Users\bruno\projects\pytest\.tmp
C:\Users\bruno\projects\pytest\.env310\Scripts\pytest.exe
C:\Users\bruno\AppData\Local\Programs\Python\Python310\python310.zip
...
```

(Note `C:\Users\bruno\projects\pytest\.env310\Scripts\pytest.exe` being in `sys.path` in the second call)

So I don't think this is architecture/OS specific.

**Comment by pmeier** [comment]
This is also not an issue from the `pytest` "binary" that is autogenerated:

```
❯ cat $(which pytest)
#!/home/philip/.conda/envs/nebari-dev/bin/python3.8
# -*- coding: utf-8 -*-
import re
import sys
from pytest import console_main
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(console_main())
```

https://github.com/pytest-dev/pytest/blob/61f7c27ec03099c0ed9c80b74fd37687e610ffa7/src/pytest/__main__.py#L1-L5
